### DISCUSSION TITLE
Merging standard and streaming data sources in panel configurations

### DISCUSSION THREAD

**Discussion by wz2b** [grafana/grafana#37909] [Platform]
Streaming/live datasources are great for single-value displays and gauges, but when I'm showing data as a time series panel I keep finding myself wishing I had a way to backfill the data from some TSDB.  For my custom datasources that's easy enough to do, but I think it would be nice to be able to do these backfills in a more standard way that lets me make use of the weather of TSDB datasource plugins already available in Grafana.

I'm not entirely sure how I would envision doing this, but as an example, I am toying around with the idea of making a live datasource for Apache Sparkplug for display IoT/IIoT data from sensors and PLCs.  In this case live data will come in Sparkplug format over an MQTT connection, but the backfill data will come from Prometheus, InfluxDB, or something else.  So what I think I want is a way to link two datasources together so that a single query response - let's call it "A" - can define both.  I would apply the rule "ignore the live/streaming component unless stopTime is about now()".

One way you might be able to do this is to modify the panels themselves - like the time series graph - so that the merging takes place there.  If you did that, you would have to define a panel with a mixed datasource where query "A" comes from the tsdb and "B" gives you the exact same data/format but comes from the live streaming datasource.  I think that's messy, though, because it will now only work for those panels that understand the special meaning of multiple responses.

The main thing is that if we were to develop a library of live datasources - as examples, MQTT, Sparkplug, BACnet, and OPC/UA - you might want to pair those with any TSDB of your choosing, and not have your only option be to have a single datasource that knows about both the live interface and the TSDB you happen to have paired together.

If there is an approach to this already and I just don't know what it is please let me know.  If there isn't yet, what is the current thinking on this?

**Comment by wz2b** [comment]
Here's one idea I came up with ... you could define a standard query in the query editor, but add an additional tab to configure the live/streaming portion, then allow them to be pointed to two different datasources.  That would let this happen with, I think, fairly minimal user interface work (maybe)
![image](https://user-images.githubusercontent.com/1521209/129458669-56f4d085-69d1-4273-a865-5790e30c5526.png)

**Comment by ryantxu** [comment]
Yes!  I love the idea/goal, but we have a ton of details to sort out.  There are a few related concepts that need to be sorted out.  Currently everything is tied to a data query, but we may:
* want a different data source for "historic" vs "live" data
* run different queries (against different data sources) for different time resolutions.  
  * "last 6 months" 
  * "last 6 hours" 
  * "right now"

See https://github.com/grafana/grafana/pull/19331 -- would like to try something similar again

**Reply by 0xMihir** [reply]
We could allow the mixed data source to be split by time stamp, so query A takes effect from t1 to t2, and B from t2 to t3, et cetera.

**Comment by iPelle** [comment]
Excellent post. I have exactly this issue. Any progress with the issue?

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
